File: scenarios/ui-health/api/internal/testutil/modeltest/traces_test.go
package modeltest_test

import (
	"testing"

	"ui-health/internal/testutil/modeltest"

	"github.com/stretchr/testify/require"
)

func TestValidateTraces_AcceptsMatchingTraces(t *testing.T) {
	errs := modeltest.ValidateTraces([]modeltest.Trace[status, event]{
		{
			Name:    "happy_path",
			Initial: statusIdle,
			Steps: []modeltest.TraceStep[status, event]{
				{Name: "start", Event: eventStart, Want: statusBusy},
				{Name: "finish", Event: eventFinish, Want: statusDone},
			},
		},
		{
			Name:    "reject_finish_before_start",
			Initial: statusIdle,
			Steps: []modeltest.TraceStep[status, event]{
				{Name: "finish", Event: eventFinish, Want: statusIdle, WantErr: true},
			},
		},
	}, transition)

	require.Empty(t, errs)
}

func TestValidateTraces_RejectsDrift(t *testing.T) {
	errs := modeltest.ValidateTraces([]modeltest.Trace[status, event]{
		{
			Name:    "bad_expectations",
			Initial: statusIdle,
			Steps: []modeltest.TraceStep[status, event]{
				{Name: "start", Event: eventStart, Want: statusDone},
				{Name: "start_again", Event: eventStart, Want: statusDone},
			},
		},
	}, transition)

	require.NotEmpty(t, errs)
	require.Contains(t, joined(errs), `bad_expectations/start: got status busy, want done`)
	require.Contains(t, joined(errs), `bad_expectations/start_again: unexpected error`)
}

File: scenarios/react-component-library/ui/src/test-utils/modeltest/traces.test.ts
import { describe, expect, it } from "vitest";

import type { WorkflowTransition } from "./matrix";
import { replayTraces, type Trace, validateTraces } from "./traces";

type State = "idle" | "busy" | "done";
type Event = "start" | "finish";

const transition: WorkflowTransition<State, Event> = (state, event) => {
  if (state === "idle" && event === "start") {
    return "busy";
  }
  if (state === "busy" && event === "finish") {
    return "done";
  }
  throw new Error("illegal transition");
};

describe("modeltest trace replay", () => {
  it("accepts matching traces", () => {
    const traces = [
      {
        name: "happy_path",
        initial: "idle",
        steps: [
          { name: "start", event: "start", want: "busy" },
          { name: "finish", event: "finish", want: "done" },
        ],
      },
      {
        name: "reject_finish_before_start",
        initial: "idle",
        steps: [{ name: "finish", event: "finish", want: "idle", wantError: true }],
      },
    ] as const satisfies readonly Trace<State, Event>[];

    expect(validateTraces(traces, transition)).toEqual([]);
    expect(() => replayTraces(traces, transition)).not.toThrow();
  });

  it("rejects drift", () => {
    const traces = [
      {
        name: "bad_expectations",
        initial: "idle",
        steps: [
          { name: "start", event: "start", want: "done" },
          { name: "start_again", event: "start", want: "done" },
        ],
      },
    ] as const satisfies readonly Trace<State, Event>[];

    const errors = validateTraces(traces, transition);

    expect(errors.join("\n")).toContain("bad_expectations/start: got state busy, want done");
    expect(errors.join("\n")).toContain("bad_expectations/start_again: unexpected error");
  });
});
